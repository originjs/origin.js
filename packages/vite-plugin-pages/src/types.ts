export type PluginOptions = {
  /**
   * current workspace directory
   * @default process.cwd()
   */
  root: string

  /**
   * directory for pages components
   * @default 'src/pages'
   */
  pagesDir: string

  /**
   * directory for layouts components
   * @default 'src/layouts'
   */
  layoutsDir: string

  /**
   * Valid file extensions for pages components
   * @default ['vue']
   */
  extensions: string[]

  /**
   * List of components to exclude
   */
  excludes?: string[]
}

export type Page = {
  /**
   * file path from pages directory
   */
  pathFromPagesDir: string

  /**
   * file path from root directory
   */
  pathFromRootDir: string

  /**
   * normalize pathFromRootDir, use to check file path and sort
   * 1. make sure all path starts with /
   * 2. convert path like /_xx, /_$xx to /xx
   * 3. convert path like /_ to /~ (use for sort, because ASCII of ~ is 126)
   */
  pathFromPagesDirNormalized: string
}

export type Route = {
  name?: string

  path: string

  component: string

  children?: Route[]

  meta?: Record<string, unknown>
}
