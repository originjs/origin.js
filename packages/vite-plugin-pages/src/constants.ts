export const MODULE_NAME = 'plugin-pages'
export const CATCH_ALL_ROUTE_PATH = '/:pathMatch(.*)*'
