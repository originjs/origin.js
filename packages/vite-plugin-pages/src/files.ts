import * as globby from 'globby'
import { replaceWithSlash } from './utils'

function getExcludes(excludes: string[]) {
  return ['node_modules', '.git', '**/__*__/**', ...excludes]
}

export function getFiles(
  directory: string,
  extensions: string[],
  excludes?: string[],
): string[] {
  // setup excludes
  excludes = getExcludes((excludes || []).map(ex => replaceWithSlash(ex)))

  // return read files by extensions
  const pattern =
    extensions.length > 1
      ? `**/*.{${extensions.join(',')}}`
      : `**/*.${extensions[0]}`
  return globby.sync(pattern, {
    gitignore: true,
    ignore: excludes,
    cwd: replaceWithSlash(directory),
    onlyFiles: true,
  })
}

/**
 * sort files:
 * 1. treat xx, _xx, _$xx as same name
 * 2. in each directory: index.vue is the first, _.vue is the last
 * @param filePaths
 */
export function sortFilePaths(filePaths: string[]): string[] {
  function normalStr(str: string): string {
    if (!str.startsWith('/')) {
      str = `/${str}`
    }
    return str
      .replace('index.vue', '') // remove 'index.vue'
      .replace('_.vue', '~.vue') // '_.vue' => '~.vue'
      .replace(/\/_/g, '/') // '/_' => '/'
      .replace(/\/\$/g, '/') // '/$' => '/'
  }

  filePaths.sort((a, b) => {
    a = normalStr(a)
    b = normalStr(b)
    if (a < b) {
      return -1
    }
    if (a > b) {
      return 1
    }
    return 0
  })
  return filePaths
}
