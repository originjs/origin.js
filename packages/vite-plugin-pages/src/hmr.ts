import { ViteDevServer } from 'vite'
import { MODULE_NAME } from './constants'

function getModule(server:ViteDevServer){
  const{ moduleGraph } =server
  const module=moduleGraph.getModuleById(MODULE_NAME)
  if (module){
    moduleGraph.invalidateModule(module)
    return module
  }
  return
}

export function handleHmr(server: ViteDevServer,RouteClear:()=> void){
  const { ws,watcher }=server
  function reloadRoutes(){
    getModule(server)
    RouteClear()
    ws.send({
      type:'full-reload',
    })
    return
  }

  watcher.on('add',async(file)=>{
    console.log('add',file)
    reloadRoutes()
  })

  watcher.on('unlink',async(file)=>{
    reloadRoutes()
  })

  watcher.on('change',async(file)=>{
    reloadRoutes()
  })
}
