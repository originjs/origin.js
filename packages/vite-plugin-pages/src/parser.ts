import { parse as parseSFC, SFCBlock } from '@vue/compiler-sfc'
import fs from 'fs'
import { join, parse } from 'path'
import { getFiles } from './files'
import { PluginOptions, Route } from './types'
import { replaceWithSlash } from './utils'
import { CATCH_ALL_ROUTE_PATH } from './constants'

/**
 * get the layout property from pages
 *
 * @param path file path in pages directory
 */
export function getLayoutProperties(path: string) {
  let layout = ''
  const source = fs.readFileSync(path, 'utf8')
  const customBlocks: SFCBlock[] = parseSFC(source).descriptor.customBlocks
  const layoutBlock = customBlocks.find(
    (node: SFCBlock) => node.type === 'layout',
  )
  if (!layoutBlock) {
    return layout
  }
  const content: string = layoutBlock.content.trim()
  const pairs = content.split(':')
  if (pairs.length == 2 && pairs[0].trim() === 'layout') {
    layout = pairs[1].trim()
  } else {
    console.error('layout property was set in wrong way!')
  }
  return layout
}

function setLayoutChildren(route: Route, options: PluginOptions, layouts: any) {
  const hasDefault = hasDefaultLayout(<string>options.layoutsDir)
  const outRoute = route
  if (route.children && route.children.length > 0) {
    outRoute.children = setLayout(route.children, options)
  }

  // When there is no layout property and no default layout, don't transform route
  if (
    (!route.meta?.layout && !hasDefault) ||
    route.path === CATCH_ALL_ROUTE_PATH
  ) {
    return outRoute
  }
  if (outRoute.meta?.layout && layouts.get(outRoute.meta?.layout)) {
    return {
      path: outRoute.path,
      component: layouts.get(outRoute.meta?.layout),
      children: [outRoute],
    }
  } else {
    return {
      path: outRoute.path,
      component: layouts.get('default'),
      children: [outRoute],
    }
  }
}

export function setLayout(routes: Route[], options: PluginOptions) {
  const layouts = getLayoutMap(
    options.layoutsDir,
    options.extensions,
    options.excludes,
  )
  return routes.map(route => setLayoutChildren(route, options, layouts))
}

function hasDefaultLayout(layoutDir: string): boolean {
  const layoutFiles = getFiles(layoutDir, ['vue'])
  for (const file of layoutFiles) {
    if (file === 'default.vue') {
      return true
    }
  }
  return false
}

function getLayoutMap(
  directory: string | undefined,
  extensions: string[] | undefined,
  excludes?: string[],
) {
  const layoutMap = new Map()
  if (!directory || !extensions) {
    return layoutMap
  }
  const layoutFiles = getFiles(directory, extensions, excludes)

  for (const file of layoutFiles) {
    const parsedFile = parse(file)
    const layoutPath = replaceWithSlash(join(directory, file))
    layoutMap.set(parsedFile.name, `() => import('/${layoutPath}')`)
  }

  return layoutMap
}
