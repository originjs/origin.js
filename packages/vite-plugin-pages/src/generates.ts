import { getLayoutProperties, setLayout } from './parser'
import { PluginOptions, Route, Page } from './types'
import { CATCH_ALL_ROUTE_PATH } from './constants'

/**
 * Return whether the current route node name represents a dynamic route
 * @param routeNodeName: route node name
 */
function isDynamicNodeName(routeNodeName: string): boolean {
  return routeNodeName.startsWith('_')
}

/**
 * Return whether the current route node name represents a nested route
 * @param routeNodeName: route node name
 */
function isNestedNodeName(routeNodeName: string): boolean {
  return routeNodeName.startsWith('$')
}

/**
 * Return whether the current route node name represents a dynamic and nested route
 * @param routeNodeName: route node name
 */
function isDynamicNestedNodeName(routeNodeName: string): boolean {
  return routeNodeName.startsWith('_$')
}

function findParentRoute(
  parentPath: string,
  routes: Route[] | undefined,
): Route | undefined {
  if (!routes || routes.length < 1) {
    return
  }
  for (const route of routes) {
    if (route.path === parentPath) {
      return route
    }
    const child = findParentRoute(parentPath, route.children)
    if (child) {
      return child
    }
  }
}

export function generateRoutes(pages: Page[], options: PluginOptions): Route[] {
  let routes: Route[] = []
  for (const page of pages) {
    const { pathFromPagesDir, pathFromRootDir } = page

    const route: Route = {
      path: '',
      component: `() => import('/${page.pathFromRootDir}')`,
      meta: { layout: getLayoutProperties(pathFromRootDir) },
    }

    const routeNodeNames = pathFromPagesDir.split('.')[0].split('/')
    let lastNodeName = routeNodeNames.pop()
    if (lastNodeName === 'index') {
      // treat xx/index as xx/, update lastNodeName
      lastNodeName = routeNodeNames.pop()
    } else if (lastNodeName === '_') {
      // treat /_ as catch all route
      lastNodeName = CATCH_ALL_ROUTE_PATH.slice(1)
    }

    // special case for 'index.vue' => '/'
    if (!lastNodeName) {
      route.path = '/'
      routes.push(route)
      continue
    }

    // setup parent route path
    for (const nodeName of routeNodeNames) {
      if (isDynamicNodeName(nodeName)) {
        route.path += `/:${nodeName.slice(1)}`
      } else if (isNestedNodeName(nodeName)) {
        route.path += `/${nodeName.slice(1)}`
      } else if (isDynamicNestedNodeName(nodeName)) {
        route.path += `/:${nodeName.slice(2)}`
      } else {
        route.path += `/${nodeName}`
      }
    }

    const isDynamic = isDynamicNodeName(lastNodeName)
    const isNested = isNestedNodeName(lastNodeName)
    const isDynamicNested = isDynamicNestedNodeName(lastNodeName)
    if (isNested || isDynamicNested) {
      const parentRoute = findParentRoute(route.path || '/', routes)
      if (parentRoute) {
        if (isNested) {
          route.path += `/${lastNodeName.slice(1)}`
        } else if (isDynamicNested) {
          route.path += `/:${lastNodeName.slice(2)}`
        }
        parentRoute.children = parentRoute.children || []
        parentRoute.children.push(route)
      } else {
        console.log('vite-plugin-pages error: can not find parent route')
      }
    } else if (isDynamic) {
      route.path += `/:${lastNodeName.slice(1)}`
      routes.push(route)
    } else {
      route.path += `/${lastNodeName}`
      routes.push(route)
    }
  }

  // find and only keep first catch all route, move it to last
  const catchAllRoute = routes.find(
    route => route.path === CATCH_ALL_ROUTE_PATH,
  )
  if (catchAllRoute) {
    routes = routes.filter(route => route.path !== CATCH_ALL_ROUTE_PATH)
    routes.push(catchAllRoute)
  }

  // set layouts to routes
  routes = setLayout(routes, options)

  return routes
}

export function generateCode(routes: Route[]): string {
  // to string, and make sure component is a function
  const str = JSON.stringify(routes).replace(
    /"component":("(.*?)")/g,
    (str, replaceFrom, replaceTo) => {
      return str.replace(replaceFrom, replaceTo)
    },
  )
  // return generated code
  return `const routes = ${str};\n export default routes;`
}
