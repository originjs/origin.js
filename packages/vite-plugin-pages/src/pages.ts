import { join } from 'path'
import { Page, PageRoutePath, Pages } from './types'
import { getFiles } from './files'
import { replaceWithSlash } from './utils'
import { CATCH_ALL_ROUTE_PATH } from './constants'
import { getLayoutProperties } from './parser'

/**
 * Return whether the current route node name represents a dynamic route
 * @param routeNodeName: route node name
 */
function isDynamicNodeName(routeNodeName: string): boolean {
  return routeNodeName.startsWith('_')
}

/**
 * Return whether the current route node name represents a nested route
 * @param routeNodeName: route node name
 */
function isNestedNodeName(routeNodeName: string): boolean {
  return routeNodeName.startsWith('$')
}

/**
 * Return whether the current route node name represents a dynamic and nested route
 * @param routeNodeName: route node name
 */
function isDynamicNestedNodeName(routeNodeName: string): boolean {
  return routeNodeName.startsWith('_$')
}

function getPageRoutePath(pathFromPagesDir: string): PageRoutePath {
  const routeNodeNames = pathFromPagesDir.split('.')[0].split('/')

  let parentRoutePath = ''
  let lastNodeName = routeNodeNames.pop()
  if (lastNodeName === 'index') {
    // treat xx/index as xx/, update lastNodeName
    lastNodeName = routeNodeNames.pop()
  } else if (lastNodeName === '_') {
    // treat /_ as catch all route
    lastNodeName = CATCH_ALL_ROUTE_PATH.slice(1)
  }

  // special case for 'index.vue' => '/'
  if (!lastNodeName) {
    return {
      parentRoutePath: '/',
      routePath: '/',
      nestedRoute: false,
    }
  }

  // setup parent route path
  for (const nodeName of routeNodeNames) {
    if (isDynamicNodeName(nodeName)) {
      parentRoutePath += `/:${nodeName.slice(1)}`
    } else if (isNestedNodeName(nodeName)) {
      parentRoutePath += `/${nodeName.slice(1)}`
    } else if (isDynamicNestedNodeName(nodeName)) {
      parentRoutePath += `/:${nodeName.slice(2)}`
    } else {
      parentRoutePath += `/${nodeName}`
    }
  }

  // setup route type by last node name
  const isDynamic = isDynamicNodeName(lastNodeName)
  const isNested = isNestedNodeName(lastNodeName)
  const isDynamicNested = isDynamicNestedNodeName(lastNodeName)

  // setup route path
  let routePath = parentRoutePath
  if (isDynamic) {
    routePath += `/:${lastNodeName.slice(1)}`
  } else if (isNested) {
    routePath += `/${lastNodeName.slice(1)}`
  } else if (isDynamicNested) {
    routePath += `/:${lastNodeName.slice(2)}`
  } else {
    routePath += `/${lastNodeName}`
  }

  return {
    parentRoutePath: parentRoutePath || '/', // if parentRoutePath is '', means parentRoutePath is /
    routePath,
    nestedRoute: isNested || isDynamicNested,
  }
}

/**
 * generate Page Object
 * @param pagesDir pages directory path
 * @param pathFromPagesDir page file path from pages directory path
 */
function getPage(pagesDir: string, pathFromPagesDir: string): Page {
  /**
   * normalize pathFromRootDir, use to check file path and sort
   * 1. make sure all path starts with /
   * 2. convert path like /_xx, /_$xx to /xx
   * 3. convert path like /_ to /~ (use for sort, because ASCII of ~ is 126)
   * @param str pathFromPagesDir
   */
  function normalized(str: string): string {
    if (!str.startsWith('/')) {
      str = `/${str}`
    }
    return str
      .replace('index.vue', '') // remove 'index.vue'
      .replace('_.vue', '~.vue') // '_.vue' => '~.vue'
      .replace(/\/_/g, '/') // '/_' => '/'
      .replace(/\/\$/g, '/') // '/$' => '/'
  }

  const pathFromRootDir = replaceWithSlash(join(pagesDir, pathFromPagesDir))
  const pathFromPagesDirNormalized = normalized(pathFromPagesDir)
  const { parentRoutePath, routePath, nestedRoute } =
    getPageRoutePath(pathFromPagesDir)
  const layout = getLayoutProperties(pathFromRootDir)

  return {
    pathFromPagesDir: pathFromPagesDir,
    pathFromRootDir,
    pathFromPagesDirNormalized,
    parentRoutePath,
    routePath,
    nestedRoute,
    layout,
  }
}

/**
 * sort pages by pathFromPagesDirNormalized:
 * in each directory: index.vue is the first, _.vue is the last
 */
export function sortPages(pages: Page[]): Page[] {
  pages.sort((a, b) => {
    const strA = a.pathFromPagesDirNormalized
    const strB = b.pathFromPagesDirNormalized
    if (strA < strB) {
      return -1
    }
    if (strA > strB) {
      return 1
    }
    return 0
  })
  return pages
}

export const pages: Pages = {
  pathSet: new Set<string>(),
  sortedPages: [],
}

export function isDuplicatedPage(page: Page): boolean {
  if (pages.pathSet.has(page.pathFromPagesDirNormalized)) {
    console.log(
      `vite-plugin-pages warn: path of ${page.pathFromPagesDir} is Duplicated`,
    )
    return true
  }
  return false
}

export function initPages(pagesDir: string, extensions: string[]): Pages {
  // clear pages
  pages.pathSet = new Set<string>()
  pages.sortedPages = []

  // init pages
  const files = getFiles(pagesDir, extensions)
  for (const file of files) {
    const page = getPage(pagesDir, file)
    if (!isDuplicatedPage(page)) {
      pages.pathSet.add(page.pathFromPagesDirNormalized)
      pages.sortedPages.push(page)
    }
  }
  pages.sortedPages = sortPages(pages.sortedPages) // sort

  return pages
}

export function updatePages(
  event: string,
  pagesDir: string,
  pathFromPagesDir: string,
) {
  // don't need to update before we call initPages
  if (pages.sortedPages.length === 0) {
    return
  }

  if (event === 'add') {
    const page = getPage(pagesDir, pathFromPagesDir)
    if (!isDuplicatedPage(page)) {
      pages.pathSet.add(page.pathFromPagesDirNormalized)
      pages.sortedPages.push(page)
      sortPages(pages.sortedPages)
    }
  } else if (event === 'unlink') {
    // find page and remove. sortedPages is still sorted after we remove
    for (let i = 0; i < pages.sortedPages.length; i++) {
      const page = pages.sortedPages[i]
      if (page.pathFromPagesDir === pathFromPagesDir) {
        pages.sortedPages = pages.sortedPages.slice(i, 1)
        pages.pathSet.delete(page.pathFromPagesDirNormalized)
        break
      }
    }
  } else if (event === 'change') {
    // update page
    for (let i = 0; i < pages.sortedPages.length; i++) {
      if (pages.sortedPages[i].pathFromPagesDir === pathFromPagesDir) {
        pages.sortedPages[i] = getPage(pagesDir, pathFromPagesDir)
        break
      }
    }
  }
}
