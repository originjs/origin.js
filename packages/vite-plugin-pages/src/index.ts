import { PluginOptions } from './types'
import { Plugin } from 'vite'
import { MODULE_NAME } from './constants'
import { generateCode, generateRoutes } from './generates'
import { getPages } from './pages'

export default (
  userOptions: PluginOptions = {
    root: process.cwd(),
    pagesDir: 'src/pages',
    layoutsDir: 'src/layouts',
    extensions: ['vue'],
  },
): Plugin => {
  const options: PluginOptions = Object.assign({}, userOptions)

  return {
    name: 'vite:pages',
    enforce: 'pre',
    resolveId(id) {
      if (id === MODULE_NAME) {
        return id
      }
      return null
    },
    async load(id) {
      if (id !== MODULE_NAME) {
        return
      }
      const pages = getPages(options.pagesDir, options.extensions)
      const routes = generateRoutes(pages, options)
      return generateCode(routes)
    },
    async transform(_code, id) {
      if (!/vue&type=layout/.test(id)) {
        return
      }
      return {
        code: 'export default {}',
        map: null,
      }
    },
  }
}

export { PluginOptions }
