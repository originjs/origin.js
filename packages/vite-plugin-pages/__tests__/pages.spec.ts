import { getPages, sortPages } from '../src/pages'

describe('pages', () => {
  const extensions = ['vue']
  const pagesDir = 'packages/vite-plugin-pages/__tests__/assets/pages'
  const winPagesDir = pagesDir.replace(/\//g, '\\')

  test('getPages', () => {
    expect(sortPages(getPages(pagesDir, extensions))).toMatchSnapshot()
    expect(sortPages(getPages(winPagesDir, extensions))).toMatchSnapshot()
  })
})
