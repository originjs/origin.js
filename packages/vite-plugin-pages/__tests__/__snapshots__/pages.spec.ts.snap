// Vitest Snapshot v1

exports[`pages > getPages 1`] = `
[
  {
    "layout": "",
    "nestedRoute": false,
    "parentRoutePath": "/",
    "pathFromPagesDir": "index.vue",
    "pathFromPagesDirNormalized": "/",
    "pathFromRootDir": "packages/vite-plugin-pages/__tests__/assets/pages/index.vue",
    "routePath": "/",
  },
  {
    "layout": "content",
    "nestedRoute": true,
    "parentRoutePath": "/",
    "pathFromPagesDir": "_$dn.vue",
    "pathFromPagesDirNormalized": "/dn.vue",
    "pathFromRootDir": "packages/vite-plugin-pages/__tests__/assets/pages/_$dn.vue",
    "routePath": "/:dn",
  },
  {
    "layout": "",
    "nestedRoute": false,
    "parentRoutePath": "/",
    "pathFromPagesDir": "user/index.vue",
    "pathFromPagesDirNormalized": "/user/",
    "pathFromRootDir": "packages/vite-plugin-pages/__tests__/assets/pages/user/index.vue",
    "routePath": "/user",
  },
  {
    "layout": "",
    "nestedRoute": false,
    "parentRoutePath": "/user",
    "pathFromPagesDir": "user/foo.vue",
    "pathFromPagesDirNormalized": "/user/foo.vue",
    "pathFromRootDir": "packages/vite-plugin-pages/__tests__/assets/pages/user/foo.vue",
    "routePath": "/user/foo",
  },
  {
    "layout": "bar",
    "nestedRoute": false,
    "parentRoutePath": "/user",
    "pathFromPagesDir": "user/_id.vue",
    "pathFromPagesDirNormalized": "/user/id.vue",
    "pathFromRootDir": "packages/vite-plugin-pages/__tests__/assets/pages/user/_id.vue",
    "routePath": "/user/:id",
  },
  {
    "layout": "",
    "nestedRoute": true,
    "parentRoutePath": "/user",
    "pathFromPagesDir": "user/$nest_user/index.vue",
    "pathFromPagesDirNormalized": "/user/nest_user/",
    "pathFromRootDir": "packages/vite-plugin-pages/__tests__/assets/pages/user/$nest_user/index.vue",
    "routePath": "/user/nest_user",
  },
  {
    "layout": "",
    "nestedRoute": false,
    "parentRoutePath": "/user/nest_user",
    "pathFromPagesDir": "user/$nest_user/foo.vue",
    "pathFromPagesDirNormalized": "/user/nest_user/foo.vue",
    "pathFromRootDir": "packages/vite-plugin-pages/__tests__/assets/pages/user/$nest_user/foo.vue",
    "routePath": "/user/nest_user/foo",
  },
  {
    "layout": "content",
    "nestedRoute": true,
    "parentRoutePath": "/user/nest_user",
    "pathFromPagesDir": "user/$nest_user/$zoom.vue",
    "pathFromPagesDirNormalized": "/user/nest_user/zoom.vue",
    "pathFromRootDir": "packages/vite-plugin-pages/__tests__/assets/pages/user/$nest_user/$zoom.vue",
    "routePath": "/user/nest_user/zoom",
  },
  {
    "layout": "default",
    "nestedRoute": false,
    "parentRoutePath": "/user/nest_user",
    "pathFromPagesDir": "user/$nest_user/_.vue",
    "pathFromPagesDirNormalized": "/user/nest_user/~.vue",
    "pathFromRootDir": "packages/vite-plugin-pages/__tests__/assets/pages/user/$nest_user/_.vue",
    "routePath": "/user/nest_user/:pathMatch(.*)*",
  },
  {
    "layout": "",
    "nestedRoute": false,
    "parentRoutePath": "/user/not_nest_user",
    "pathFromPagesDir": "user/not_nest_user/foo.vue",
    "pathFromPagesDirNormalized": "/user/not_nest_user/foo.vue",
    "pathFromRootDir": "packages/vite-plugin-pages/__tests__/assets/pages/user/not_nest_user/foo.vue",
    "routePath": "/user/not_nest_user/foo",
  },
  {
    "layout": "default",
    "nestedRoute": false,
    "parentRoutePath": "/user/not_nest_user",
    "pathFromPagesDir": "user/not_nest_user/_.vue",
    "pathFromPagesDirNormalized": "/user/not_nest_user/~.vue",
    "pathFromRootDir": "packages/vite-plugin-pages/__tests__/assets/pages/user/not_nest_user/_.vue",
    "routePath": "/user/not_nest_user/:pathMatch(.*)*",
  },
  {
    "layout": "content",
    "nestedRoute": false,
    "parentRoutePath": "/user",
    "pathFromPagesDir": "user/zoom.vue",
    "pathFromPagesDirNormalized": "/user/zoom.vue",
    "pathFromRootDir": "packages/vite-plugin-pages/__tests__/assets/pages/user/zoom.vue",
    "routePath": "/user/zoom",
  },
  {
    "layout": "content",
    "nestedRoute": true,
    "parentRoutePath": "/",
    "pathFromPagesDir": "$zoom.vue",
    "pathFromPagesDirNormalized": "/zoom.vue",
    "pathFromRootDir": "packages/vite-plugin-pages/__tests__/assets/pages/$zoom.vue",
    "routePath": "/zoom",
  },
  {
    "layout": "default",
    "nestedRoute": false,
    "parentRoutePath": "/",
    "pathFromPagesDir": "_.vue",
    "pathFromPagesDirNormalized": "/~.vue",
    "pathFromRootDir": "packages/vite-plugin-pages/__tests__/assets/pages/_.vue",
    "routePath": "/:pathMatch(.*)*",
  },
]
`;

exports[`pages > getPages 2`] = `
[
  {
    "layout": "",
    "nestedRoute": false,
    "parentRoutePath": "/",
    "pathFromPagesDir": "index.vue",
    "pathFromPagesDirNormalized": "/",
    "pathFromRootDir": "packages/vite-plugin-pages/__tests__/assets/pages/index.vue",
    "routePath": "/",
  },
  {
    "layout": "content",
    "nestedRoute": true,
    "parentRoutePath": "/",
    "pathFromPagesDir": "_$dn.vue",
    "pathFromPagesDirNormalized": "/dn.vue",
    "pathFromRootDir": "packages/vite-plugin-pages/__tests__/assets/pages/_$dn.vue",
    "routePath": "/:dn",
  },
  {
    "layout": "",
    "nestedRoute": false,
    "parentRoutePath": "/",
    "pathFromPagesDir": "user/index.vue",
    "pathFromPagesDirNormalized": "/user/",
    "pathFromRootDir": "packages/vite-plugin-pages/__tests__/assets/pages/user/index.vue",
    "routePath": "/user",
  },
  {
    "layout": "",
    "nestedRoute": false,
    "parentRoutePath": "/user",
    "pathFromPagesDir": "user/foo.vue",
    "pathFromPagesDirNormalized": "/user/foo.vue",
    "pathFromRootDir": "packages/vite-plugin-pages/__tests__/assets/pages/user/foo.vue",
    "routePath": "/user/foo",
  },
  {
    "layout": "bar",
    "nestedRoute": false,
    "parentRoutePath": "/user",
    "pathFromPagesDir": "user/_id.vue",
    "pathFromPagesDirNormalized": "/user/id.vue",
    "pathFromRootDir": "packages/vite-plugin-pages/__tests__/assets/pages/user/_id.vue",
    "routePath": "/user/:id",
  },
  {
    "layout": "",
    "nestedRoute": true,
    "parentRoutePath": "/user",
    "pathFromPagesDir": "user/$nest_user/index.vue",
    "pathFromPagesDirNormalized": "/user/nest_user/",
    "pathFromRootDir": "packages/vite-plugin-pages/__tests__/assets/pages/user/$nest_user/index.vue",
    "routePath": "/user/nest_user",
  },
  {
    "layout": "",
    "nestedRoute": false,
    "parentRoutePath": "/user/nest_user",
    "pathFromPagesDir": "user/$nest_user/foo.vue",
    "pathFromPagesDirNormalized": "/user/nest_user/foo.vue",
    "pathFromRootDir": "packages/vite-plugin-pages/__tests__/assets/pages/user/$nest_user/foo.vue",
    "routePath": "/user/nest_user/foo",
  },
  {
    "layout": "content",
    "nestedRoute": true,
    "parentRoutePath": "/user/nest_user",
    "pathFromPagesDir": "user/$nest_user/$zoom.vue",
    "pathFromPagesDirNormalized": "/user/nest_user/zoom.vue",
    "pathFromRootDir": "packages/vite-plugin-pages/__tests__/assets/pages/user/$nest_user/$zoom.vue",
    "routePath": "/user/nest_user/zoom",
  },
  {
    "layout": "default",
    "nestedRoute": false,
    "parentRoutePath": "/user/nest_user",
    "pathFromPagesDir": "user/$nest_user/_.vue",
    "pathFromPagesDirNormalized": "/user/nest_user/~.vue",
    "pathFromRootDir": "packages/vite-plugin-pages/__tests__/assets/pages/user/$nest_user/_.vue",
    "routePath": "/user/nest_user/:pathMatch(.*)*",
  },
  {
    "layout": "",
    "nestedRoute": false,
    "parentRoutePath": "/user/not_nest_user",
    "pathFromPagesDir": "user/not_nest_user/foo.vue",
    "pathFromPagesDirNormalized": "/user/not_nest_user/foo.vue",
    "pathFromRootDir": "packages/vite-plugin-pages/__tests__/assets/pages/user/not_nest_user/foo.vue",
    "routePath": "/user/not_nest_user/foo",
  },
  {
    "layout": "default",
    "nestedRoute": false,
    "parentRoutePath": "/user/not_nest_user",
    "pathFromPagesDir": "user/not_nest_user/_.vue",
    "pathFromPagesDirNormalized": "/user/not_nest_user/~.vue",
    "pathFromRootDir": "packages/vite-plugin-pages/__tests__/assets/pages/user/not_nest_user/_.vue",
    "routePath": "/user/not_nest_user/:pathMatch(.*)*",
  },
  {
    "layout": "content",
    "nestedRoute": false,
    "parentRoutePath": "/user",
    "pathFromPagesDir": "user/zoom.vue",
    "pathFromPagesDirNormalized": "/user/zoom.vue",
    "pathFromRootDir": "packages/vite-plugin-pages/__tests__/assets/pages/user/zoom.vue",
    "routePath": "/user/zoom",
  },
  {
    "layout": "content",
    "nestedRoute": true,
    "parentRoutePath": "/",
    "pathFromPagesDir": "$zoom.vue",
    "pathFromPagesDirNormalized": "/zoom.vue",
    "pathFromRootDir": "packages/vite-plugin-pages/__tests__/assets/pages/$zoom.vue",
    "routePath": "/zoom",
  },
  {
    "layout": "default",
    "nestedRoute": false,
    "parentRoutePath": "/",
    "pathFromPagesDir": "_.vue",
    "pathFromPagesDirNormalized": "/~.vue",
    "pathFromRootDir": "packages/vite-plugin-pages/__tests__/assets/pages/_.vue",
    "routePath": "/:pathMatch(.*)*",
  },
]
`;
